### PR TITLE
Feature/18 auth

### PR DESCRIPTION
**Pull request by VladimirNee**
add:
redux slice and thunk for auth logic
screens for auth logic + placeholder for email validation
additional types and styles

### CODE REVIEW

File: src/components/ProfileHeader/ProfileHeader.tsx [hunk 1]
```diff
@@ -32,14 +33,22 @@ const ProfileHeader = ({ isLoggedIn }: Props) => {
 
           <View style={styles.redirectAuthContainer}>
             <Button
-              title="Signin"
+              title="Sign In"
               onPress={() => {
                 console.log('redirect to signin!');
```
Comment (javokhirm) [inline src/components/ProfileHeader/ProfileHeader.tsx]: remove console logs

File: package.json [hunk 2]
```diff
@@ -36,7 +36,9 @@
     "react": "18.2.0",
     "react-native": "0.72.6",
     "react-redux": "^8.1.3",
-    "redux-persist": "^6.0.0"
+    "redux-persist": "^6.0.0",
+    "react-native-web": "~0.19.6",
```
Comment (javokhirm) [inline package.json]: why do you need `react-native-web`?

File: src/screens/auth/Signin/Signin.tsx [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+import styles from '../auth.styles';
+
+const Signin = () => {
+  const navigation = useNavigation<NavigationProp<RootStackParamList>>();
+  const dispatch = useAppDispatch();
+  const loading = useSelector(getAuthLoading);
+  const authError = useSelector(getAuthError);
+
+  const [validationErrors, setValidationErrors] = useState<Record<string, string>>({});
+  const [formInteracted, setFormInteracted] = useState<boolean>(false);
```
Comment (javokhirm) [inline src/screens/auth/Signin/Signin.tsx:23]: you don't need `formInteracted` state

File: src/screens/auth/Signin/Signin.tsx [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+          style={styles.input}
+          placeholder="Enter your password"
+          secureTextEntry
+          value={credentials.password}
+          onChangeText={(text) => setCredentials({ ...credentials, password: text })}
+        />
+        <Seperator />
+        <View style={styles.container}>
+          <Text style={styles.toggleText}>Don't have an account yet?</Text>
+          <Text
```
Comment (javokhirm) [inline src/screens/auth/Signin/Signin.tsx:91]: style `Sign Up` text, it should be diplayed as it's a link

File: src/screens/auth/Signup/Signup.tsx [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+  useEffect(() => {
+    dispatch(authActions.clearError());
+  }, []);
+
+  return (
+    <ScreenTemplate>
+      <FormTemplate
+        loading={loading}
+        onSubmit={handleSignup}
+        error={formInteracted && formIsValid ? authError : null}
```
Comment (javokhirm) [inline src/screens/auth/Signup/Signup.tsx]: authError message is enough here, but keep in mind you should pass an error prop as string

File: src/screens/auth/Signin/Signin.tsx [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        <Input
+          style={styles.input}
+          value={credentials.email}
+          onChangeText={(text) => setCredentials({ ...credentials, email: text })}
+          placeholder="Enter your email"
+          error={validationErrors.email}
+        />
+        <Input
+          style={styles.input}
+          placeholder="Enter your password"
```
Comment (javokhirm) [inline src/screens/auth/Signin/Signin.tsx:84]: inputs have the same color for placeholder and text, let's make them different

File: src/screens/auth/Signup/Signup.tsx [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+
+import { validateForm } from './Signup.utils';
+import styles from '../auth.styles';
+
+const Signup = () => {
+  const navigation = useNavigation<NavigationProp<RootStackParamList>>();
+
+  const dispatch = useAppDispatch();
+  const loading = useSelector(getAuthLoading);
+  const authError = useSelector(getAuthError);
```
Comment (javokhirm) [inline src/screens/auth/Signup/Signup.tsx]: clear auth error when the page is opened and after the form submission

File: src/store/slices/authSlice.ts [hunk 8]
```diff
@@ -0,0 +1,65 @@
+import AsyncStorage from '@react-native-async-storage/async-storage';
+import { createSlice } from '@reduxjs/toolkit';
+import { AuthState } from 'src/types';
+
+import { AsyncThunks } from '../thunks';
+
+const initialState: AuthState = {
+  id: '',
+  tokens: null,
```
Comment (javokhirm) [inline src/store/slices/authSlice.ts]: don't store tokens on reducer, better use `expo-secure-store`

File: src/store/slices/authSlice.ts [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+  },
+  extraReducers: (builder) => {
+    // Sign in reducers
+    builder.addCase(AsyncThunks.signIn.pending, (state) => {
+      state.pending = true;
+    });
+    builder.addCase(AsyncThunks.signIn.fulfilled, (state, action) => {
+      state.pending = false;
+      state.id = action.payload.id;
+      AsyncStorage.setItem('tokens', JSON.stringify(action.payload.tokens));
```
Comment (javokhirm) [inline src/store/slices/authSlice.ts]: use a secure store to save sensitive info (tokens)

File: src/screens/auth/Signup/Signup.tsx [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+  };
+
+  const handleInputChange = (field: string, value: string) => {
+    setCredentials({
+      ...credentials,
+      [field]: value,
+    });
+  };
+
+  const isPasswordMatch = credentials.password === credentials.confirm_password;
```
Comment (javokhirm) [inline src/screens/auth/Signup/Signup.tsx]: let's check this on `validateForm` function

File: src/screens/auth/Signin/Signin.tsx [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+  useEffect(() => {
+    dispatch(authActions.clearError());
+  }, []);
+
+  return (
+    <ScreenTemplate>
+      <FormTemplate
+        loading={loading}
+        onSubmit={handleSignin}
+        error={formInteracted && formIsValid ? authError : null}
```
Comment (javokhirm) [inline src/screens/auth/Signin/Signin.tsx]: authError message is enough

File: src/screens/auth/Signup/Signup.tsx [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+
+  const formIsValid = Object.keys(validationErrors).length === 0;
+
+  const handleSignup = () => {
+    setFormInteracted(true);
+    if (formIsValid) {
+      console.log('Signing up...');
+      dispatch(AsyncThunks.signUp(credentials));
+      console.log('redirecting to verify email');
+      navigation.navigate('VerifyEmail');
```
Comment (javokhirm) [inline src/screens/auth/Signup/Signup.tsx]: navigate only if server returns success

**Review by javokhirm — CHANGES_REQUESTED** [review]

File: package.json (hunk 2 above)
Comment (VladimirNee) [reply package.json]: removed locally & will update with next commit

File: src/screens/auth/Signin/Signin.tsx (hunk 3 above)
Comment (VladimirNee) [reply src/screens/auth/Signin/Signin.tsx:23]: removed

**Comment by VladimirNee** [comment]
resolved according to comments, but excluding styling of inputs. This needs to be discussed with Khushnud Meliev

**Review by javokhirm — APPROVED** [review]

**Pull request merged**